### PR TITLE
GAWB-479: remove workflows from listSubmissions; make only one query

### PR DESCRIPTION
**Pull request by davidangb**
This PR removes workflow information from listSubmissions, and now retrieves this list of submissions in a single query instead of a loooot of queries.

This does not require a corresponding UI change, since I'm updating (with PO approval) an existing endpoint.

Question for someone knowledgable: the previous code seems to support a case where a submission does not have an associated entity. Is this even possible? My code (probably) does not work for this case.

I fear that once we land this change, and suddenly we are able to navigate the list of submissions, we will find that drilling into the detail of a single submission is too slow. That's outside the scope of this PR.

@dvoet tech lead ... @jmthibault79 you won the wheel a second time in a row. Please feel free to reassign to someone else or back to me if you don't have the bandwidth.
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Make sure liquibase is updated if appropriate
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to develop 
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by dvoet** [comment]
re: "Question for someone knowledgable: the previous code seems to support a case where a submission does not have an associated entity. Is this even possible? My code (probably) does not work for this case."

This is not currently possible due to FK constraints but was possible in the orient days. This will become possible again when the delete story hits but any fixes to this code to support delete should happen then

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala [hunk 1]
```diff
@@ -100,11 +100,19 @@ trait SubmissionComponent {
...
         userRec <- rawlsUserQuery if (submissionRec.submitterId === userRec.userSubjectId)
-      } yield (submissionRec, userRec)
+        methodConfigRec <- methodConfigurationQuery if (submissionRec.methodConfigurationId === methodConfigRec.id)
+        entityRec <- entityQuery if (submissionRec.submissionEntityId === entityRec.id)
+      } yield (submissionRec, userRec, methodConfigRec, entityRec)
       
-      query.result.flatMap(recs => DBIO.sequence(recs.map { case (submissionRec, userRec) =>
-        loadSubmission(submissionRec.id).map(sub => (sub.get, rawlsUserQuery.unmarshalRawlsUser(userRec)))
-      }))
+      query.result.flatMap{recs => DBIO.sequence(recs.map {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala]: do you need the DBIO.sequence? I think by removing DBIO.sequence and DBIO.successful below you don't need flatMap and just put this code in the yield of the for comprehension.

**Comment by dvoet** [comment]
1 minor comment then :+1:

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala (hunk 1 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala]: Agreed.  I was able to make it work with `query.result.map` and removal of both `DBIO`s.

Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala]: I haven't been able to get this to work. Inside the for-yield, those vars represent the slick Table classes ... it's only after the query.result mapping that they represent the individual rows within those tables. But, I was able to get rid of the DBIO.sequence and DBIO.successful by using map, not flatMap ... see the third commit in this PR.

**Comment by davidangb** [comment]
gotcha and thank you on the submission:entity constraint.

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 2]
```diff
@@ -952,7 +952,7 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: SlickDataSo
     dataSource.inTransaction { dataAccess =>
       withWorkspaceContextAndPermissions(workspaceName, WorkspaceAccessLevels.Read, dataAccess) { workspaceContext =>
         dataAccess.submissionQuery.listWithSubmitter(workspaceContext).map { _.map {
-            case (submission, user) => new SubmissionStatusResponse(submission, user) 
+            case (submission, user) => new SubmissionListResponse(submission, user)
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: If `listWithSubmitter` returned a `ReadAction[Seq[SubmissionListResponse]]` directly, this map step would not be necessary.

File: src/main/resources/swagger/rawls.yaml [hunk 3]
```diff
@@ -743,7 +743,7 @@ paths:
           schema:
             type: array
             items:
-              $ref: '#/definitions/Submission'
+              $ref: '#/definitions/SubmissionList'
```
Comment (jmthibault79) [inline src/main/resources/swagger/rawls.yaml]: SubmissionList_Response_

File: src/main/resources/swagger/rawls.yaml [hunk 4]
```diff
@@ -2906,6 +2906,35 @@ definitions:
...
+      submitter:
+        type: string
+        description: user that created the submission
+      methodConfigurationNamespace:
+        type: string
+        description: Method configuration namespace
+      methodConfigurationName:
+        type: string
+        description: Method configuration name
+      entityType:
```
Comment (jmthibault79) [inline src/main/resources/swagger/rawls.yaml]: `submissionEntity: AttributeEntityReference` - one JSON level down.

**Comment by jmthibault79** [comment]
A few small comments.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala (hunk 1 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala]: New version looks good to me.

**Comment by davidangb** [comment]
updated to reflect all the good PR feedback.

File: src/main/resources/swagger/rawls.yaml [hunk 5]
```diff
@@ -3289,6 +3286,19 @@ definitions:
         type: ''
         description: Map[String,Attribute]
 
+  AttributeEntityReference:
+    description: ''
+    required:
+      - entityName
```
Comment (jmthibault79) [inline src/main/resources/swagger/rawls.yaml]: Nitpick: will output these in the reverse order so ideally the docs should match

**Comment by jmthibault79** [comment]
:+1: with nitpick.  Squash and merge after tests pass.

**Pull request merged**